### PR TITLE
Adding Mongodb element to add `arrayMatches` the #4762

### PR DESCRIPTION
**Pull request by jeremypiednoel**

### CODE REVIEW

**Comment by jeremypiednoel** [comment]
All DB tests are in `spec/DatabaseController.spec.js` ?

**Comment by flovilmart** [comment]
Some are there, also an edge to edge test. I still am not able to see how the current features prevent you doing what you expect :)

**Comment by jeremypiednoel** [comment]
here is my workarroud 

```js
  const objects = [1,2,3,4,5,6,7,8].map((idx) => {
  const obj = new Parse.Object('Object');
  obj.set('key', idx);
  return obj;
});
  let parent3;
  Parse.Object.saveAll(objects).then(() => {

    const parent = new Parse.Object('Parent');
    parent.set('objects', objects.slice(0, 3));

    const shift = objects.shift();
    const parent2 = new Parse.Object('Parent');
    parent2.set('objects', [objects[1],shift]);

    parent3 = new Parse.Object('Parent');
    parent3.set('objects', objects.slice(1, 4));

    return Parse.Object.saveAll([parent, parent2, parent3]);
  }).then(() => {
    const query = new Parse.Query('Parent');
    query["_where"] = {
      $nor : [{ objects : { $elemMatch : { $nin : objects.map(object => object.toPointer()) } }}]
    };
    return query.find();
  }).then((result) => {
    expect(parent3.id).toEqual(result[0].id);
    expect(result.length).toBe(1);
  }).then(done).catch(done.fail);
```

**Comment by flovilmart** [comment]
Not sure why you need the `$nor` operator as you provide only one element. $elemMatch should be enough no?

**Comment by jeremypiednoel** [comment]
The condition is a `nin` because de `in` return at least one i revert the test.
So the logic is 
1 - find me all documents where at least one array element is *not* on our list
2 - negate the entire query

**Comment by jeremypiednoel** [comment]
i made a lot of research and i found that : http://www.askasya.com/post/matchallarrayelements/ where the guy is doing exactly that.

**Comment by jeremypiednoel** [comment]
I also found this case where they use double negation as well.
I think it's more a missing feature in mongo

**Comment by flovilmart** [comment]
So it's more `containsExclusivelySome(key, array)`:

```
I want to fetch all objects where object's key contents contains **only** values provided in the array.
```

**Comment by jeremypiednoel** [comment]
Do you still think it worth it to add it ? or it's a specific need that i should keep private ?

**Comment by flovilmart** [comment]
Let's see if the test fails, as they should because those operators are not supported on postgresql.

**Comment by jeremypiednoel** [comment]
it looks like tests succeeded

**Comment by flovilmart** [comment]
I prefer not to have both supports separated as we may forget to implement it or worst be unable to provide a cross DB support

**Comment by flovilmart** [comment]
Let’s see how it goes for this one, but the operator chain look very specific to mongo, and perhaps we can create a parse specific operator that also work with Postgres :)

**Comment by dplewis** [comment]
@jeremypiednoel Can you write a few edge tests so we can play with? Postgres has `any` / `some` that you can use for array comparison.

https://www.postgresql.org/docs/9.4/static/functions-comparisons.html

**Comment by jeremypiednoel** [comment]
hey @dplewis,
Thanks for the answer.
Do you have an exemple in the parse spec of what kind of edges would you like ?

**Comment by dplewis** [comment]
@jeremypiednoel Test cases using the REST API is what I mean. 

This PR https://github.com/parse-community/parse-server/pull/3864 is a good starting place as it is an inverse to your issue. You can start by replacing `$all` with `$elemMatch` and `$nor`.

**Comment by jeremypiednoel** [comment]
Ok i'll do that. Thanks @dplewis and @flovilmart.

**Comment by jeremypiednoel** [comment]
@flovilmart @dplewis i added the edge test.
Sorry for the delay.
Please tell me if it's enough or if you guys need more.

File: spec/ParseQuery.spec.js [hunk 1]
```diff
@@ -523,7 +523,7 @@ describe('Parse.Query testing', () => {
     Parse.Object.saveAll(objectList).then((results) => {
       equal(objectList.length, results.length);
 
-      return require('request-promise').get({
+      return rp.get({
```
Comment (dplewis) [inline spec/ParseQuery.spec.js]: We typically look for changes related to the PR. But the clean up here is OK.

**Comment by dplewis** [comment]
@jeremypiednoel I have this working for Postgres. As @flovilmart stated your query is mongo specific so we'll introduce a new parse specific operator. I'll call it `$containedBy` for now after the postgres `<@`contained by operator. (subject to change)

If its ok with you I can commit my changes directly to this branch.

**Comment by flovilmart** [comment]
Yep 👍

**Comment by dplewis** [comment]
@jeremypiednoel How does this look?

**Comment by jeremypiednoel** [comment]
@dplewis it looks very good ! thanks for that.

**Comment by jeremypiednoel** [comment]
I can update the parse JS API and iOS SDK if you guys want with the `containedBy` method

**Comment by flovilmart** [comment]
That would be very neat :)

**Comment by jeremypiednoel** [comment]
Should we precise on the doc that this feature will be available only with the X.X version of parse server ?

**Comment by dplewis** [comment]
@flovilmart how does this look?

**Comment by flovilmart** [comment]
Depends if we have the equivalent easily with PG :)

**Comment by jeremypiednoel** [comment]
I might be wrong but I think it would be a combinaison of NOT and OR in postgresql. @dplewis  ?

**Comment by dplewis** [comment]
@jeremypiednoel Feel free to open a separate PR with tests. That’s what I’m thinking for PG.

**Comment by jeremypiednoel** [comment]
Ok i will !
Should i first add support into JS and iOS client ?

**Comment by dplewis** [comment]
Let’s get this merged first. I’ll do the PHP and Android SDK. We still have to wait for the next release tho

**Comment by flovilmart** [comment]
@dplewis I can put a release together now :)

**Comment by dplewis** [comment]
@jeremypiednoel Thanks for getting started on this. I’m still surprise this isn’t supported in mongo out of the box. I’ll open feature request on the mongo repo.

@flovilmart Thanks.

**Comment by jeremypiednoel** [comment]
Yes i was surprised to but the documentation is pretty clear : its either `$elemMatch` ,`$all` or `$size` :  https://docs.mongodb.com/manual/reference/operator/query-array/

**Comment by flovilmart** [comment]
this is very weird!

**Pull request merged**